tls/extensions/s2n_client_key_share.c

```diff
@@ -56,27 +57,92 @@ const s2n_extension_type s2n_client_key_share_extension = {
     .if_missing = s2n_extension_noop_if_missing,
 };
 
-static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+static int s2n_generate_preferred_key_shares(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
-    notnull_check(conn);
-    notnull_check(conn->config);
+    uint8_t key_shares = conn->preferred_key_shares;
+    const struct s2n_ecc_named_curve *named_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
 
     const struct s2n_ecc_preferences *ecc_pref = NULL;
     GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
     notnull_check(ecc_pref);
 
+    bool lsb_set = key_shares & 1;
+
+    for (int i = 1; i <= ecc_pref->count; i++) {
+        ecc_evp_params = &conn->secure.client_ecc_evp_params[i-1];
+        named_curve = ecc_pref->ecc_curves[i-1];
+        ecc_evp_params->negotiated_curve = named_curve;
+        ecc_evp_params->evp_pkey = NULL;
+        /* If lsb is set, skip keyshare generation for all curve */
+        if (lsb_set) {
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->iana_id));
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->share_size));
+            uint8_t *data = s2n_stuffer_raw_write(out, ecc_evp_params->negotiated_curve->share_size);
+            *data = 0;
+        } else if ((key_shares >> i) & 1) { /* If bit other than lsb is set, generate keyshare for the corresponding curve */
```

Comment:
You might want to wrap this check in a named macro to make it more intuitive even without the comment. Like "KEYSHARE_IS_REQUESTED" or something. I'd also rename "lsb_set" to something like "empty_keyshares"